src/cpp/session/modules/jobs/JobsApi.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * JobsApi.cpp
+ *
+ * Copyright (C) 2009-18 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <session/SessionModuleContext.hpp>
+ 
+#include "JobsApi.hpp"
+
+enum JobUpdateType
+{
+   JobAdded   = 0,
+   JobUpdated = 1,
+   JobRemoved = 2
+};
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {      
+namespace jobs {
+
+namespace {
+
+// map of job ID to jobs
+std::map<std::string, boost::shared_ptr<Job> > s_jobs;
```

Comment:
Is it okay to have a complex object like this as a static object? I recall we had RStudio crashes-on-exit when attempting to destruct maps like this in the past; it might be better just to explicitly `new` and leak this.

Reply:
For context: https://github.com/rstudio/rstudio/blob/9a2cd883d53a98b7b6c0957d6fc492e673ee5aac/src/cpp/session/SessionRpc.cpp#L297-L302

Reply:
That's a good question -- this code just moved from a different file, where it's been in the dailies/previews for a while now without issue. Presumably if this were a problem we'd have seen it by now?

Reply:
I think you're right -- if we've been running safely with it for a while then I think no change is needed.